_posts/2019-03-26-gloo-onion-layers.md

```diff
@@ -0,0 +1,250 @@
+    });
+
+// Spawn our future to run it!
+wasm_bindgen_futures::spawn_local(my_future);
+```
+
+Note that we use `futures` 0.1 for now, because we've fought tooth and nail to
+get the Wasm ecosystem on stable Rust, but as soon as the new
+`std::future::Future` design is stable, we plan to switch over. We are very
+excited for `async`/`await` as well!
+
+## More Layers?
+
+That's all the layers we have for the `setTimeout` and `setInterval`
+APIs. Different Web APIs will have different sets of layers, and this is
+fine. Not every Web API uses callbacks, so it doesn't make sense to always have
+a `callback` module in every Gloo crate. The important part is that we are
+actively identifying layers, making them public and reusable, and building
+higher-level layers on top of lower-level layers.
+
+We will likely add even higher-level layers to other Web APIs where it makes
+sense. For example, the [File API][]'s `FileReader` interface exposes methods
+that you shouldn't call until after certain events have fired, and any attempt
+to call them earlier will throw. We can codify this as [a state machine-based
+`Future`][state-machine-future], that doesn't even give you the ability to call
+those methods until after the relevant events have fired and the state machine
+reaches a certain state. Leveraging types at compile time for ergonomics and
+correctness!
+
+Another future direction is adding more integration layers with more parts of
+the larger Rust crates ecosystem. For example, adding functional reactive
+programming-style layers via [the `futures-signals`
+crate][integrate-futures-signals] that the [`dominator`][dominator] framework is
+built upon.
+
+## Events
+
+One of the active bits of design work going on in Gloo right now is how to craft
+our event targets and listeners layer. Events are used across most of the Web
+APIs, so it is very important we get this design right, as it will sit
+underneath many of our other crates. While we haven't 100% nailed down the
+design yet, I really like where we are headed.
+
+On top of [`web_sys::Event`][web-sys-event] and
+[`web_sys::EventTarget::add_event_listener_with_callback`][web-sys-add-listener],
+we are building a layer for [adding and removing event
+listeners][raii-listeners] and managing their lifetimes from Rust via RAII-style
+automatic cleanup upon drop. This will enable usage like this:
+
+```rust
+use gloo::events::EventListener;
+
+// Get an event target from somewhere. Maybe a DOM element or maybe the window
+// or document.
+let target: web_sys::EventTarget = unimplemented!();
+
+let listener = EventListener::new(&target, "click", |event: web_sys::Event| {
+    // Cast the `Event` into a `MouseEvent`
+    let mouse_event = event.dyn_into::<web_sys::MouseEvent>().unwrap();
+
+    // Do stuff on click...
+});
+
+// If we want to remove the listener, we just drop it:
+drop(listener);
+
+// Alternatively, if we want to listen forever, we can use `forget`:
+listener.forget();
+```
```

Comment:
I think it's really important that we explain the *idiomatic* use of `EventListener`.

Right now it's just showing `drop` and `forget`, both of which are *highly* unidiomatic: `drop` will cause the event to never fire, and `forget` leaks memory!

Instead, the primary use case of `EventListener` is to store it in a struct, so that way the event listener will be tied to the lifetime of the struct:

```rust
use futures::stream::Stream;
use futures::sync::mpsc::{unbounded, UnboundedReceiver};

struct OnClick {
    receiver: UnboundedReceiver<web_sys::MouseEvent>,
    _on_click: EventListener,
}

impl OnClick {
    #[inline]
    pub fn new(target: &web_sys::EventTarget) -> Self {
        let (sender, receiver) = unbounded();

        Self {
            receiver,
            _on_click: EventListener::new(target, "click", move |event| {
                let event = event.dyn_into::<web_sys::MouseEvent>().unwrap_throw();

                sender.unbounded_send(event).unwrap_throw();
            }),
        }
    }
}

impl Stream for OnClick {
    type Item = web_sys::MouseEvent;
    type Error = ();

    #[inline]
    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
        self.receiver.poll()
    }
}
```

Now when the `OnClick` stream is dropped/cancelled, it will automatically cleanup the `EventListener` and the `Closure`.